content/learning-paths/cross-platform/loop-reflowing/autovectorization-on-arm-1.md

```diff
@@ -0,0 +1,227 @@
+
+This is quite nice and you can expect a performance gain of about 4x faster in general.
+
+You might be wondering if there is a way to hint to the compiler that the sizes are always going to be multiples of 4 and avoid the last part of the code and the answer is *yes*, but it depends on the compiler. In the case of gcc, it is enough to add an instruction that ensures the size is only multiples of 4:
+
+```C
+int32_t dotprod(int32_t *A, int32_t *B, size_t N) {
+    int32_t result = 0;
+    N -= N % 4;
+    for (size_t i=0; i < N; i++) {
+        result += A[i]*B[i];
+    }
+    return result;
+}
+```
+
+And the assembly output with `-O3` will much more compact, now that it does not have to handle the left over bytes:
+
+```as
+dotprod:
+        ands    x2, x2, -4
+        beq     .L4
+        movi    v0.4s, 0
+        lsl     x3, x2, 2
+        mov     x2, 0
+.L3:
+        ldr     q2, [x1, x2]
+        ldr     q1, [x0, x2]
+        add     x2, x2, 16
+        mla     v0.4s, v2.4s, v1.4s
+        cmp     x3, x2
+        bne     .L3
+        addv    s0, v0.4s
+        fmov    w0, s0
+        ret
+.L4:
+        fmov    s0, wzr
+        fmov    w0, s0
+        ret
+```
+
+But is that all that the compiler can do? Thankfully not. Modern compilers are very proficient in generating code that utilizes all available instructions, provided they have the right information.
+For example, the `dotprod()` function that you wrote operates on `int32_t` elements, what if you could limit the range to 8-bit? Something like that is not untypical, and we know there is an Armv8 ISA extension that [provides a pair of new `SDOT`/`UDOT` instructions to perform a dotprot across 8-bit elements of 2 vectors and store the results in the 32-bit elements of the resulting vector](https://developer.arm.com/documentation/102651/a/What-are-dot-product-intructions-). Could the compiler make use of that automatically? Or should you resort to a hand-written version that uses the respective intrinsics?
+
+It turns out that some compilers can, and will detect that the number or the iterations is a multiple of the number of elements in a SIMD vector. Convert the code to use `int8_t` types for `A` and `B` arrays:
+
+```C
+int32_t dotprod(int8_t *A, int8_t *B, size_t N) {
+    int32_t result = 0;
+    N -= N % 4;
+    for (size_t i=0; i < N; i++) {
+        result += A[i]*B[i];
+    }
+    return result;
+}
+
+int main() {
+    const int N = 128;
+    int8_t A[N], B[N];
+
+    int32_t dot = dotprod(A, B, N);
+
+    printf("dotprod = %d", dot);
+}
+```
+
+You need to recompile the code with `gcc -O3 -Wall -g -fno-inline -march=armv8-a+dotprod` in order to hint to the compiler that it has the new instructions at its disposal.
```

Comment:
it should mention `-march=armv8-a+dotprod` separately because the sentence currently gives the impression that the `-Wall -g` which are new here, are also required.

Reply:
agreed, will fix this